File: frontend/mobile/lib/core/services/prayer_calendar_service.dart
import 'package:dio/dio.dart';
import '../network/dio_client.dart';
import '../network/api_endpoints.dart';
import '../../features/prayer_calendar/data/models/calendar_day_model.dart';

/// Prayer Calendar Service for monthly prayer times
class PrayerCalendarService {
  final DioClient _dioClient;

  PrayerCalendarService(this._dioClient);

  /// Get monthly prayer times calendar
  Future<MonthlyCalendarModel> getMonthlyCalendar({
    required double latitude,
    required double longitude,
    required int hijriYear,
    required int hijriMonth,
    String? calculationMethod,
  }) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.prayerCalendar}/$hijriYear/$hijriMonth',
        queryParameters: {
          'latitude': latitude,
          'longitude': longitude,
          if (calculationMethod != null) 'method': calculationMethod,
        },
      );

      return MonthlyCalendarModel.fromJson(response.data['data']);
    } catch (e) {
      throw Exception('Failed to fetch monthly calendar: $e');
    }
  }

  /// Get prayer times for a specific date range
  Future<List<CalendarDayModel>> getPrayerTimesRange({
    required double latitude,
    required double longitude,
    required DateTime startDate,
    required DateTime endDate,
    String? calculationMethod,
  }) async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.prayerTimesRange,
        queryParameters: {
          'latitude': latitude,
          'longitude': longitude,
          'start_date': startDate.toIso8601String().split('T')[0],
          'end_date': endDate.toIso8601String().split('T')[0],
          if (calculationMethod != null) 'method': calculationMethod,
        },
      );

      return (response.data['data'] as List)
          .map((json) => CalendarDayModel.fromJson(json))
          .toList();
    } catch (e) {
      throw Exception('Failed to fetch prayer times range: $e');
    }
  }

  /// Get Islamic events for a specific month
  Future<List<IslamicEventModel>> getIslamicEvents({
    int? hijriMonth,
    int? hijriYear,
    int? importanceLevel,
  }) async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.islamicEvents,
        queryParameters: {
          if (hijriMonth != null) 'hijri_month': hijriMonth,
          if (hijriYear != null) 'hijri_year': hijriYear,
          if (importanceLevel != null) 'importance_level': importanceLevel,
        },
      );

      return (response.data['data'] as List)
          .map((json) => IslamicEventModel.fromJson(json))
          .toList();
    } catch (e) {
      throw Exception('Failed to fetch Islamic events: $e');
    }
  }

  /// Export calendar to iCal format
  Future<String> exportCalendarToICal({
    required double latitude,
    required double longitude,
    required int hijriYear,
    required int hijriMonth,
  }) async {
    try {
      final response = await _dioClient.get(
        '${ApiEndpoints.prayerCalendar}/$hijriYear/$hijriMonth/export',
        queryParameters: {
          'latitude': latitude,
          'longitude': longitude,
          'format': 'ical',
        },
        options: Options(responseType: ResponseType.plain),
      );

      return response.data as String;
    } catch (e) {
      throw Exception('Failed to export calendar: $e');
    }
  }

  /// Get shareable calendar link
  Future<String> getShareableLink({
    required double latitude,
    required double longitude,
    required int hijriYear,
    required int hijriMonth,
  }) async {
    try {
      final response = await _dioClient.post(
        '${ApiEndpoints.prayerCalendar}/share',
        data: {
          'latitude': latitude,
          'longitude': longitude,
          'hijri_year': hijriYear,
          'hijri_month': hijriMonth,
        },
      );

      return response.data['data']['share_url'] as String;
    } catch (e) {
      throw Exception('Failed to generate shareable link: $e');
    }
  }
}

File: frontend/nextjs-app/src/lib/services/prayer-calendar-service.ts
import { axiosClient } from '../api/axios-client';
import { API_ENDPOINTS } from '../api/endpoints';
import type {
  MonthlyCalendar,
  CalendarDay,
  IslamicEvent,
  CalendarExportOptions,
} from '@/types/prayer-calendar';

export class PrayerCalendarService {
  /**
   * Get monthly prayer times calendar
   */
  static async getMonthlyCalendar(
    latitude: number,
    longitude: number,
    hijriYear: number,
    hijriMonth: number,
    calculationMethod?: string
  ): Promise<MonthlyCalendar> {
    const response = await axiosClient.get<{ data: MonthlyCalendar }>(
      `${API_ENDPOINTS.PRAYER_CALENDAR}/${hijriYear}/${hijriMonth}`,
      {
        params: {
          latitude,
          longitude,
          ...(calculationMethod && { method: calculationMethod }),
        },
      }
    );
    return response.data.data;
  }

  /**
   * Get prayer times for a date range
   */
  static async getPrayerTimesRange(
    latitude: number,
    longitude: number,
    startDate: string,
    endDate: string,
    calculationMethod?: string
  ): Promise<CalendarDay[]> {
    const response = await axiosClient.get<{ data: CalendarDay[] }>(
      API_ENDPOINTS.PRAYER_TIMES_RANGE,
      {
        params: {
          latitude,
          longitude,
          start_date: startDate,
          end_date: endDate,
          ...(calculationMethod && { method: calculationMethod }),
        },
      }
    );
    return response.data.data;
  }

  /**
   * Get Islamic events for a specific month
   */
  static async getIslamicEvents(
    hijriMonth?: number,
    hijriYear?: number,
    importanceLevel?: number
  ): Promise<IslamicEvent[]> {
    const response = await axiosClient.get<{ data: IslamicEvent[] }>(
      API_ENDPOINTS.ISLAMIC_EVENTS,
      {
        params: {
          ...(hijriMonth && { hijri_month: hijriMonth }),
          ...(hijriYear && { hijri_year: hijriYear }),
          ...(importanceLevel && { importance_level: importanceLevel }),
        },
      }
    );
    return response.data.data;
  }

  /**
   * Export calendar to iCal format
   */
  static async exportCalendarToICal(
    latitude: number,
    longitude: number,
    hijriYear: number,
    hijriMonth: number
  ): Promise<string> {
    const response = await axiosClient.get<string>(
      `${API_ENDPOINTS.PRAYER_CALENDAR}/${hijriYear}/${hijriMonth}/export`,
      {
        params: {
          latitude,
          longitude,
          format: 'ical',
        },
        responseType: 'text',
      }
    );
    return response.data;
  }

  /**
   * Get shareable calendar link
   */
  static async getShareableLink(
    latitude: number,
    longitude: number,
    hijriYear: number,
    hijriMonth: number
  ): Promise<string> {
    const response = await axiosClient.post<{ data: { share_url: string } }>(
      `${API_ENDPOINTS.PRAYER_CALENDAR}/share`,
      {
        latitude,
        longitude,
        hijri_year: hijriYear,
        hijri_month: hijriMonth,
      }
    );
    return response.data.data.share_url;
  }

  /**
   * Download calendar as file
   */
  static async downloadCalendar(
    calendar: MonthlyCalendar,
    options: CalendarExportOptions
  ): Promise<Blob> {
    if (options.format === 'ical') {
      // For iCal, we already have the export endpoint
      const icalData = await this.exportCalendarToICal(
        0, // These would come from the calendar context
        0,
        calendar.hijriYear,
        calendar.hijriMonth.monthNumber
      );
      return new Blob([icalData], { type: 'text/calendar' });
    }

    // For other formats, implement client-side generation
    throw new Error(`Export format ${options.format} not yet implemented`);
  }

  /**
   * Helper: Check if a day is special (Friday or has events)
   */
  static isSpecialDay(day: CalendarDay): boolean {
    return day.isFriday || day.events.length > 0;
  }

  /**
   * Helper: Get day background color based on its properties
   */
  static getDayBackgroundColor(day: CalendarDay): string {
    if (day.isToday) return 'bg-primary-50 border-accent';
    if (day.isFriday) return 'bg-secondary-50';
    if (day.events.some((e) => e.eventType === 'eid')) return 'bg-green-50';
    if (day.events.length > 0) return 'bg-accent-50';
    return 'bg-white';
  }

  /**
   * Helper: Get day text color based on its properties
   */
  static getDayTextColor(day: CalendarDay): string {
    if (day.isToday) return 'text-primary';
    if (day.isFriday) return 'text-secondary';
    if (day.events.some((e) => e.eventType === 'eid')) return 'text-green-600';
    return 'text-gray-900';
  }

  /**
   * Helper: Format Hijri date
   */
  static formatHijriDate(hijriDate: HijriDate): string {
    return `${hijriDate.weekdayArabic}، ${hijriDate.day} ${hijriDate.monthNameArabic} ${hijriDate.year} هـ`;
  }

  /**
   * Helper: Format Gregorian date
   */
  static formatGregorianDate(dateString: string): string {
    const date = new Date(dateString);
    const months = [
      'يناير',
      'فبراير',
      'مارس',
      'أبريل',
      'مايو',
      'يونيو',
      'يوليو',
      'أغسطس',
      'سبتمبر',
      'أكتوبر',
      'نوفمبر',
      'ديسمبر',
    ];
    const weekdays = [
      'الأحد',
      'الاثنين',
      'الثلاثاء',
      'الأربعاء',
      'الخميس',
      'الجمعة',
      'السبت',
    ];

    return `${weekdays[date.getDay()]}، ${date.getDate()} ${
      months[date.getMonth()]
    } ${date.getFullYear()} م`;
  }
}
